contracts/incentives/base/BaseIncentivesController.sol

```diff
@@ -0,0 +1,207 @@
+    external
+    view
+    override
+    returns (uint256)
+  {
+    uint256 unclaimedRewards = _usersUnclaimedRewards[user];
+
+    DistributionTypes.UserStakeInput[] memory userState =
+      new DistributionTypes.UserStakeInput[](assets.length);
+    for (uint256 i = 0; i < assets.length; i++) {
+      userState[i].underlyingAsset = assets[i];
+      (userState[i].stakedByUser, userState[i].totalStaked) = IScaledBalanceToken(assets[i])
+        .getScaledUserBalanceAndSupply(user);
+    }
+    unclaimedRewards = unclaimedRewards.add(_getUnclaimedRewards(user, userState));
+    return unclaimedRewards;
+  }
+
+  /// @inheritdoc IAaveIncentivesController
+  function claimRewards(
+    address[] calldata assets,
+    uint256 amount,
+    address to
+  ) external override returns (uint256) {
+    require(to != address(0), 'INVALID_TO_ADDRESS');
+    return _claimRewards(assets, amount, msg.sender, msg.sender, to);
+  }
+
+  /// @inheritdoc IAaveIncentivesController
+  function claimRewardsOnBehalf(
+    address[] calldata assets,
+    uint256 amount,
+    address user,
+    address to
+  ) external override onlyAuthorizedClaimers(msg.sender, user) returns (uint256) {
+    require(user != address(0), 'INVALID_USER_ADDRESS');
+    require(to != address(0), 'INVALID_TO_ADDRESS');
+    return _claimRewards(assets, amount, msg.sender, user, to);
+  }
+
+  /// @inheritdoc IAaveIncentivesController
+  function claimRewardsToSelf(address[] calldata assets, uint256 amount)
+    external
+    override
+    returns (uint256)
+  {
+    return _claimRewards(assets, amount, msg.sender, msg.sender, msg.sender);
+  }
+
+  /// @inheritdoc IAaveIncentivesController
+  function setClaimer(address user, address caller) external override onlyEmissionManager {
+    _authorizedClaimers[user] = caller;
+    emit ClaimerSet(user, caller);
+  }
+
+  /// @inheritdoc IAaveIncentivesController
+  function getClaimer(address user) external view override returns (address) {
+    return _authorizedClaimers[user];
+  }
+
+  /// @inheritdoc IAaveIncentivesController
+  function getUserUnclaimedRewards(address _user) external view override returns (uint256) {
+    return _usersUnclaimedRewards[_user];
+  }
+
+  /**
+   * @dev returns the revision of the implementation contract
+   */
+  function getRevision() internal pure override returns (uint256) {
+    return REVISION;
+  }
+
+  /**
+   * @dev Claims reward for an user on behalf, on all the assets of the lending pool, accumulating the pending rewards.
+   * @param amount Amount of rewards to claim
+   * @param user Address to check and claim rewards
+   * @param to Address that will be receiving the rewards
+   * @return Rewards claimed
+   **/
+  function _claimRewards(
+    address[] calldata assets,
+    uint256 amount,
+    address claimer,
+    address user,
+    address to
+  ) internal returns (uint256) {
+    if (amount == 0) {
+      return 0;
+    }
+    uint256 unclaimedRewards = _usersUnclaimedRewards[user];
```

Comment:
If `amount <= unclaimedRewards` it is possible to shortcircuit the claim and just use the unclaimed to payout the reward no need to compute how much have accrued. Expect it to save a lot of gas when the case is true, and only spend little when it is not because we already cached the `unclaimedRewards`.

Reply:
Great find! Will make this change.

Reply:
Done, let me know if it meets your criteria or if you thought another way.